File: src/main/java/com/example/MonaServer/Helper/ImageHelper.java
package com.example.MonaServer.Helper;

import org.springframework.core.io.ClassPathResource;
import org.springframework.core.io.Resource;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.*;
import java.util.Arrays;

public class ImageHelper {

    public static final int SIZE = 500;
    public static final int SIZE_PIN = 100;
    public static final int X_OFFSET = 11;
    public static final int Y_OFFSET = 4;
    public static final int DIAMETER = 79;
    public static final int HEX_COLOR_TRANSPARENT = 0xFFFFFF;


    public static byte[] getProfileImage(byte[] image) {
        try {
            BufferedImage imageBuff = resizeImage(image, SIZE);
            ByteArrayOutputStream buffer = new ByteArrayOutputStream();
            ImageIO.write(imageBuff, "png", buffer);
            imageBuff.flush();
            return buffer.toByteArray();
        } catch (IOException | IllegalStateException e) {
            throw new IllegalStateException("image does not have the right size constrains");
        }
    }

    public static byte[] getPinImage(byte[] image) {
        try {
            //scale to SIZE_PIN x SIZE_PIN
            BufferedImage imageBuff = resizeImage(image, DIAMETER);
            //get Resources
            BufferedImage pinImage = getImageFromResources("pin_image.png");
            BufferedImage pinBorder = getImageFromResources("pin_border.png");
            //create return image
            BufferedImage returnImage = new BufferedImage(SIZE_PIN, SIZE_PIN, BufferedImage.TYPE_INT_ARGB);
            Graphics2D g = returnImage.createGraphics();
            for (int x = 0; x < SIZE_PIN; x++) {
                for (int y = 0; y < SIZE_PIN; y++) {
                    if (x >= X_OFFSET &&
                            x <= X_OFFSET + DIAMETER &&
                            y >= Y_OFFSET &&
                            y <= Y_OFFSET + DIAMETER &&
                            isNotTransparent(pinImage, x, y)) {           //draw image if pixel in pin_image.png is not transparent
                        g.setColor(new Color(imageBuff.getRGB(x - X_OFFSET,y - Y_OFFSET)));
                    } else if (isNotTransparent(pinBorder, x, y)) {   //draw image if pixel in pin_border.png is not transparent
                        g.setColor(new Color(pinBorder.getRGB(x,y)));
                    } else {                                        //default transparent
                        g.setColor(new Color(0,0,0,0));
                    }
                    g.drawLine(x,y,x,y);
                }
            }
            ByteArrayOutputStream buffer = new ByteArrayOutputStream();
            ImageIO.write(returnImage, "png", buffer);
            g.dispose();
            imageBuff.flush();
            pinImage.flush();
            pinBorder.flush();
            returnImage.flush();
            return buffer.toByteArray();
        } catch (Exception e) {
            System.out.println(e);
            throw new IllegalStateException("image does not have the right size constrains");
        }
    }

    private static boolean isNotTransparent(BufferedImage image, int x, int y ) {
        int pixel = image.getRGB(x,y);
        return (pixel >> 24) != 0x00;
    }

    private static BufferedImage resizeImage(byte[] image, int size) throws IOException {
        ByteArrayInputStream in = new ByteArrayInputStream(image);
        BufferedImage img = ImageIO.read(in);
        in.close();
        Image scaledImage = img.getScaledInstance(size, size, Image.SCALE_SMOOTH);
        BufferedImage imageBuff = new BufferedImage(size, size, BufferedImage.TYPE_INT_RGB);
        imageBuff.getGraphics().drawImage(scaledImage, 0, 0, new Color(0,0,0), null);
        scaledImage.flush();
        return imageBuff;
    }

    private static BufferedImage getImageFromResources(String name) throws IOException {
        Resource resource = new ClassPathResource("pin/" + name);
        InputStream input = resource.getInputStream();
        ByteArrayInputStream in = new ByteArrayInputStream(input.readAllBytes());
        BufferedImage img = ImageIO.read(in);
        input.close();
        in.close();
        return img;
    }
}

File: src/main/kotlin/de/lrprojects/monaserver/helper/ImageHelper.kt
package de.lrprojects.monaserver.helper

import de.lrprojects.monaserver.excepetion.ImageProcessingException
import de.lrprojects.monaserver.types.ImageQualityType
import net.coobird.thumbnailator.Thumbnails
import org.springframework.core.io.ClassPathResource
import org.springframework.core.io.Resource
import org.springframework.stereotype.Component
import java.awt.Color
import java.awt.image.BufferedImage
import java.io.*
import javax.imageio.ImageIO
import kotlin.math.min


@Component
class ImageHelper {

    @Throws(ImageProcessingException::class)
    fun getProfileImage(image: ByteArray): ByteArray {
        return compressImage(image, SIZE_PROFILE, SIZE_PROFILE,true)
    }

    @Throws(ImageProcessingException::class)
    fun getProfileImageSmall(image: ByteArray): ByteArray {
        return compressImage(image, SIZE_PROFILE_SMALL, SIZE_PROFILE_SMALL, true)
    }

    @Throws(ImageProcessingException::class)
    fun getPinImage(image: ByteArray): ByteArray {
        return try {
            //scale to SIZE_PIN x SIZE_PIN
            val imageBuff = resizeImage(image, SIZE_GROUP_PIN_DIAMETER)
            //get Resources
            //create return image
            val returnImage = BufferedImage(SIZE_GROUP_PIN, SIZE_GROUP_PIN, BufferedImage.TYPE_INT_ARGB)
            val g = returnImage.createGraphics()
            for (x in 0 until SIZE_GROUP_PIN) {
                for (y in 0 until SIZE_GROUP_PIN) {
                    if (x >= SIZE_GROUP_PIN_X_OFFSET
                        && x <= SIZE_GROUP_PIN_X_OFFSET + SIZE_GROUP_PIN_DIAMETER
                        && y >= SIZE_GROUP_PIN_Y_OFFSET
                        && y <= SIZE_GROUP_PIN_Y_OFFSET + SIZE_GROUP_PIN_DIAMETER
                        && isNotTransparent(pinImage, x, y))
                    {
                        //draw image if pixel in pin_image.png is not transparent
                        g.color = Color(imageBuff.getRGB(x - SIZE_GROUP_PIN_X_OFFSET, y - SIZE_GROUP_PIN_Y_OFFSET))
                    } else if (isNotTransparent(pinBorder, x, y)) {
                        //draw image if pixel in pin_border.png is not transparent
                        g.color = Color(pinBorder.getRGB(x, y))
                    } else {
                        //default transparent
                        g.color = Color(0, 0, 0, 0)
                    }
                    g.drawLine(x, y, x, y)
                }
            }
            val buffer = ByteArrayOutputStream()
            ImageIO.write(returnImage, OUTPUT_FORMAT_PNG, buffer)
            g.dispose()
            imageBuff.flush()
            returnImage.flush()
            buffer.toByteArray()
        } catch (e: IllegalStateException) {
            throw ImageProcessingException("image does not have the right size constrains")
        } catch (e: FileNotFoundException) {
            throw ImageProcessingException("Static image template could not be accessed")
        } catch (e: NullPointerException) {
            throw ImageProcessingException("Image could not be processed, try a different format.")
        }
    }

    private fun isNotTransparent(image: BufferedImage, x: Int, y: Int): Boolean {
        val pixel = image.getRGB(x, y)
        return pixel shr 24 != 0x00
    }

    @Throws(ImageProcessingException::class)
    fun compressPinImage(image: ByteArray): ByteArray {
        return compressImage(image, WIDTH_PIN, HEIGHT_PIN)
    }

    @Throws(ImageProcessingException::class)
    private fun compressImage(image: ByteArray, width: Int, height: Int, forceSize: Boolean = false): ByteArray {
        try {
            val `in` = ByteArrayInputStream(image)
            val out = ByteArrayOutputStream()

            val originalSizeKB = image.size / 1024
            val dynamicQuality = calculateImageQuality(originalSizeKB)

            val builder = Thumbnails.of(`in`)
            if (forceSize) {
                builder.size(width, height)
            } else {
                builder.forceSize(width, height)
            }
            builder.outputQuality(dynamicQuality)
                .outputFormat(OUTPUT_FORMAT_JPG)
                .toOutputStream(out)

            `in`.close()
            val transformedImage = out.toByteArray()
            out.close()
            return transformedImage
        } catch (e: IOException) {
            throw ImageProcessingException("Image could not be compressed: ${e.message}")
        }
    }

    private fun calculateImageQuality(sizeKB: Int): Double {
        // Base the quality on size and resolution thresholds
        val sizeFactor = when {
            sizeKB > ImageQualityType.VERY_HIGH.kbSize -> ImageQualityType.VERY_HIGH.quality
            sizeKB > ImageQualityType.HIGH.kbSize -> ImageQualityType.HIGH.quality
            sizeKB > ImageQualityType.MEDIUM.kbSize -> ImageQualityType.MEDIUM.quality
            else ->  ImageQualityType.LOW.quality
        }

        // Ensure quality stays within bounds [0.1, 1.0]
        return min(sizeFactor, 1.0)
    }



    @Throws(IOException::class)
    fun resizeImage(image: ByteArray, width: Int): BufferedImage {
        val `in` = ByteArrayInputStream(image)
        val originalImage = ImageIO.read(`in`)
        `in`.close()

        val originalWidth = originalImage.width
        val originalHeight = originalImage.height

        // Calculate the proportional height based on the provided width
        val height = (width.toDouble() / originalWidth.toDouble() * originalHeight.toDouble()).toInt()

        // Create a scaled instance of the original image
        val scaledImage = originalImage.getScaledInstance(width, height, BufferedImage.SCALE_SMOOTH)

        // Create a new BufferedImage with the scaled dimensions
        val resizedImage = BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)

        // Draw the scaled image onto the new BufferedImage
        val graphics2D = resizedImage.createGraphics()
        graphics2D.drawImage(scaledImage, 0, 0, Color(0, 0, 0), null)
        graphics2D.dispose()

        return resizedImage
    }

    companion object {
        private fun getImageFromResources(name: String): BufferedImage {
            val resource: Resource = ClassPathResource("pin/$name")
            val input = resource.inputStream
            val `in` = ByteArrayInputStream(input.readAllBytes())
            val img = ImageIO.read(`in`)
            input.close()
            `in`.close()
            return img
        }

        private val pinImage: BufferedImage = getImageFromResources("pin_image.png")
        private val pinBorder: BufferedImage = getImageFromResources("pin_border.png")
        private const val SIZE_PROFILE = 500
        private const val SIZE_GROUP_PIN = 100
        private const val SIZE_PROFILE_SMALL = 100
        private const val SIZE_GROUP_PIN_X_OFFSET = 11
        private const val SIZE_GROUP_PIN_Y_OFFSET = 4
        private const val SIZE_GROUP_PIN_DIAMETER = 79
        private const val WIDTH_PIN = 720
        private const val HEIGHT_PIN = 960
        private const val OUTPUT_FORMAT_JPG = "jpg"
        private const val OUTPUT_FORMAT_PNG = "png"
    }
}